### ISSUE TITLE
Celery workers encountering deadlocks when attempting to execute queries against table djkombu_message...

### ISSUE THREAD

**Issue by seegraff** [celery/kombu#524]
Here's the result of my SHOW ENGINE INNODB STATUS query:

     2015-10-07 08:34:56 27f0
     *** (1) TRANSACTION:
     TRANSACTION 9707249, ACTIVE 0 sec starting index read
     mysql tables in use 1, locked 1
     LOCK WAIT 3 lock struct(s), heap size 1184, 2 row lock(s)
     MySQL thread id 1622, OS thread handle 0x4a4, query id 191767 localhost ::1 root Sending data
     SELECT `djkombu_message`.`id`, `djkombu_message`.`visible`, `djkombu_message`.`sent_at`, `djkombu_message`.`payload`, `djkombu_message`.`queue_id` FROM `djkombu_message` WHERE (`djkombu_message`.`queue_id` = 3 AND `djkombu_message`.`visible` = 1) ORDER BY `djkombu_message`.`sent_at` ASC, `djkombu_message`.`id` ASC LIMIT 1 FOR UPDATE
     *** (1) WAITING FOR THIS LOCK TO BE GRANTED:
     RECORD LOCKS space id 2347 page no 64 n bits 80 index `PRIMARY` of table `dotalabs`.`djkombu_message` trx id 9707249 lock_mode X locks rec but not gap waiting
     Record lock, heap no 8 PHYSICAL RECORD: n_fields 7; compact format; info bits 32
      0: len 4; hex 80000123; asc    #;;
      1: len 6; hex 000000941ee2; asc       ;;
      2: len 7; hex 5a0000023b16d4; asc Z   ;  ;;
      3: len 1; hex 80; asc  ;;
      4: len 8; hex 99974e889503a082; asc   N     ;;
      5: len 30; hex 7b2268656164657273223a207b7d2c2022636f6e74656e742d7479706522; asc {"headers": {}, "content-type"; (total 1770 bytes);
      6: len 4; hex 80000003; asc     ;;
     
     *** (2) TRANSACTION:
     TRANSACTION 9707234, ACTIVE 0 sec updating or deleting, thread declared inside InnoDB 4999
     mysql tables in use 1, locked 1
     9 lock struct(s), heap size 2936, 5 row lock(s), undo log entries 2
     MySQL thread id 1616, OS thread handle 0x27f0, query id 191778 localhost ::1 root updating
     DELETE FROM djkombu_message WHERE visible=0
     *** (2) HOLDS THE LOCK(S):
     RECORD LOCKS space id 2347 page no 64 n bits 80 index `PRIMARY` of table `dotalabs`.`djkombu_message` trx id 9707234 lock_mode X locks rec but not gap
     Record lock, heap no 8 PHYSICAL RECORD: n_fields 7; compact format; info bits 32
      0: len 4; hex 80000123; asc    #;;
      1: len 6; hex 000000941ee2; asc       ;;
      2: len 7; hex 5a0000023b16d4; asc Z   ;  ;;
      3: len 1; hex 80; asc  ;;
      4: len 8; hex 99974e889503a082; asc   N     ;;
      5: len 30; hex 7b2268656164657273223a207b7d2c2022636f6e74656e742d7479706522; asc {"headers": {}, "content-type"; (total 1770 bytes);
      6: len 4; hex 80000003; asc     ;;
     
     *** (2) WAITING FOR THIS LOCK TO BE GRANTED:
     RECORD LOCKS space id 2347 page no 7 n bits 952 index `djkombu_message_df2f2974` of table `dotalabs`.`djkombu_message` trx id 9707234 lock_mode X locks rec but not gap waiting
     Record lock, heap no 65 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
      0: len 8; hex 99974e889503a082; asc   N     ;;
      1: len 4; hex 80000123; asc    #;;
     
     *** WE ROLL BACK TRANSACTION (1)

Here's the stack trace from the exception:

    [2015-10-07 08:34:56,571: ERROR/MainProcess] Unrecoverable error: OperationalError(1213, 'Deadlock found when trying to get lock; try restarting transaction')
    Traceback (most recent call last):
      File "C:\Python34\lib\site-packages\django\db\backends\utils.py", line 64, in execute
        return self.cursor.execute(sql, params)
      File "C:\Python34\lib\site-packages\django\db\backends\mysql\base.py", line 124, in execute
        return self.cursor.execute(query, args)
      File "C:\Python34\lib\site-packages\MySQLdb\cursors.py", line 220, in execute
        self.errorhandler(self, exc, value)
      File "C:\Python34\lib\site-packages\MySQLdb\connections.py", line 36, in defaulterrorhandler
        raise errorvalue
      File "C:\Python34\lib\site-packages\MySQLdb\cursors.py", line 209, in execute
        r = self._query(query)
      File "C:\Python34\lib\site-packages\MySQLdb\cursors.py", line 371, in _query
        rowcount = self._do_query(q)
      File "C:\Python34\lib\site-packages\MySQLdb\cursors.py", line 336, in _do_query
        self._do_get_result()
      File "C:\Python34\lib\site-packages\MySQLdb\cursors.py", line 146, in _do_get_result
        self._result = self._get_result()
      File "C:\Python34\lib\site-packages\MySQLdb\cursors.py", line 368, in _get_result
        def _get_result(self): return self._get_db().store_result()
    _mysql_exceptions.OperationalError: (1213, 'Deadlock found when trying to get lock; try restarting transaction')
    
    The above exception was the direct cause of the following exception:
    
    Traceback (most recent call last):
      File "C:\Python34\lib\site-packages\celery\worker\__init__.py", line 206, in start
        self.blueprint.start(self)
      File "C:\Python34\lib\site-packages\celery\bootsteps.py", line 123, in start
        step.start(parent)
      File "C:\Python34\lib\site-packages\celery\bootsteps.py", line 374, in start
        return self.obj.start()
      File "C:\Python34\lib\site-packages\celery\worker\consumer.py", line 278, in start
        blueprint.start(self)
      File "C:\Python34\lib\site-packages\celery\bootsteps.py", line 123, in start
        step.start(parent)
      File "C:\Python34\lib\site-packages\celery\worker\consumer.py", line 821, in start
        c.loop(*c.loop_args())
      File "C:\Python34\lib\site-packages\celery\worker\loops.py", line 97, in synloop
        connection.drain_events(timeout=2.0)
      File "C:\Python34\lib\site-packages\kombu\connection.py", line 275, in drain_events
        return self.transport.drain_events(self.connection, **kwargs)
      File "C:\Python34\lib\site-packages\kombu\transport\virtual\__init__.py", line 830, in drain_events
        item, channel = get(timeout=timeout)
      File "C:\Python34\lib\site-packages\kombu\transport\virtual\scheduling.py", line 39, in get
        return self.fun(resource, **kwargs), resource
      File "C:\Python34\lib\site-packages\kombu\transport\virtual\__init__.py", line 850, in _drain_channel
        return channel.drain_events(timeout=timeout)
      File "C:\Python34\lib\site-packages\kombu\transport\virtual\__init__.py", line 642, in drain_events
        return self._poll(self.cycle, timeout=timeout)
      File "C:\Python34\lib\site-packages\kombu\transport\virtual\__init__.py", line 328, in _poll
        return cycle.get()
      File "C:\Python34\lib\site-packages\kombu\transport\virtual\scheduling.py", line 39, in get
        return self.fun(resource, **kwargs), resource
      File "C:\Python34\lib\site-packages\kombu\transport\django\__init__.py", line 49, in _get
        m = Queue.objects.fetch(queue)
      File "C:\Python34\lib\site-packages\kombu\transport\django\managers.py", line 37, in fetch
        return queue.messages.pop()
      File "C:\Python34\lib\site-packages\kombu\transport\django\managers.py", line 21, in _commit
        return fun(*args, **kwargs)
      File "C:\Python34\lib\site-packages\kombu\transport\django\managers.py", line 71, in pop
        result = resultset[0:1].get()
      File "C:\Python34\lib\site-packages\django\db\models\query.py", line 328, in get
        num = len(clone)
      File "C:\Python34\lib\site-packages\django\db\models\query.py", line 144, in __len__
        self._fetch_all()
      File "C:\Python34\lib\site-packages\django\db\models\query.py", line 965, in _fetch_all
        self._result_cache = list(self.iterator())
      File "C:\Python34\lib\site-packages\django\db\models\query.py", line 238, in iterator
        results = compiler.execute_sql()
      File "C:\Python34\lib\site-packages\django\db\models\sql\compiler.py", line 840, in execute_sql
        cursor.execute(sql, params)
      File "C:\Python34\lib\site-packages\django\db\backends\utils.py", line 79, in execute
        return super(CursorDebugWrapper, self).execute(sql, params)
      File "C:\Python34\lib\site-packages\django\db\backends\utils.py", line 64, in execute
        return self.cursor.execute(sql, params)
      File "C:\Python34\lib\site-packages\django\db\utils.py", line 97, in __exit__
        six.reraise(dj_exc_type, dj_exc_value, traceback)
      File "C:\Python34\lib\site-packages\django\utils\six.py", line 658, in reraise
        raise value.with_traceback(tb)
      File "C:\Python34\lib\site-packages\django\db\backends\utils.py", line 64, in execute
        return self.cursor.execute(sql, params)
      File "C:\Python34\lib\site-packages\django\db\backends\mysql\base.py", line 124, in execute
        return self.cursor.execute(query, args)
      File "C:\Python34\lib\site-packages\MySQLdb\cursors.py", line 220, in execute
        self.errorhandler(self, exc, value)
      File "C:\Python34\lib\site-packages\MySQLdb\connections.py", line 36, in defaulterrorhandler
        raise errorvalue
      File "C:\Python34\lib\site-packages\MySQLdb\cursors.py", line 209, in execute
        r = self._query(query)
      File "C:\Python34\lib\site-packages\MySQLdb\cursors.py", line 371, in _query
        rowcount = self._do_query(q)
      File "C:\Python34\lib\site-packages\MySQLdb\cursors.py", line 336, in _do_query
        self._do_get_result()
      File "C:\Python34\lib\site-packages\MySQLdb\cursors.py", line 146, in _do_get_result
        self._result = self._get_result()
      File "C:\Python34\lib\site-packages\MySQLdb\cursors.py", line 368, in _get_result
        def _get_result(self): return self._get_db().store_result()
    django.db.utils.OperationalError: (1213, 'Deadlock found when trying to get lock; try restarting transaction')

I'm not sure what else I can provide. I've tried troubleshooting this issue on my own but I can't seem to find any other reports of an issue like this, and it seems to be an issue with the workers encountering the deadlock when attempting to execute a query against a row in the table that's already been locked. I have two workers running, one running on a high priority queue, and the other running on a normal and low priority queues. The high priority worker should never be attempting to access queue messages that the normal/low priority worker should, so I'm not sure what else to do except report this as a bug.

**Comment by Novarg** [comment]
Same here. Got that when started to use few workers instead of one and spawn lot of async subtasks that finish very quickly. But in my case it was one queue.

Also when i'm using `READ-COMMITTED` transaction isolation i'm getting deadlocks but if i switch to default `REPEATABLE-READ` then workers die with integrity error:

```python
Traceback (most recent call last):
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/celery/worker/__init__.py", line 206, in start
    self.blueprint.start(self)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/celery/bootsteps.py", line 123, in start
    step.start(parent)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/celery/bootsteps.py", line 374, in start
    return self.obj.start()
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/celery/worker/consumer.py", line 278, in start
    blueprint.start(self)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/celery/bootsteps.py", line 123, in start
    step.start(parent)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/celery/worker/consumer.py", line 821, in start
    c.loop(*c.loop_args())
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/celery/worker/loops.py", line 97, in synloop
    connection.drain_events(timeout=2.0)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/kombu/connection.py", line 275, in drain_events
    return self.transport.drain_events(self.connection, **kwargs)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/kombu/transport/virtual/__init__.py", line 830, in drain_events
    item, channel = get(timeout=timeout)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/kombu/transport/virtual/scheduling.py", line 39, in get
    return self.fun(resource, **kwargs), resource
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/kombu/transport/virtual/__init__.py", line 850, in _drain_channel
    return channel.drain_events(timeout=timeout)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/kombu/transport/virtual/__init__.py", line 642, in drain_events
    return self._poll(self.cycle, timeout=timeout)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/kombu/transport/virtual/__init__.py", line 328, in _poll
    return cycle.get()
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/kombu/transport/virtual/scheduling.py", line 39, in get
    return self.fun(resource, **kwargs), resource
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/kombu/transport/django/__init__.py", line 49, in _get
    m = Queue.objects.fetch(queue)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/kombu/transport/django/managers.py", line 37, in fetch
    return queue.messages.pop()
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/django/db/transaction.py", line 224, in inner
    return func(*args, **kwargs)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/kombu/transport/django/managers.py", line 73, in pop
    result.save()
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/django/db/models/base.py", line 463, in save
    self.save_base(using=using, force_insert=force_insert, force_update=force_update)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/django/db/models/base.py", line 551, in save_base
    result = manager._insert([self], fields=fields, return_id=update_pk, using=using, raw=raw)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/django/db/models/manager.py", line 203, in _insert
    return insert_query(self.model, objs, fields, **kwargs)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/django/db/models/query.py", line 1593, in insert_query
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 912, in execute_sql
    cursor.execute(sql, params)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/django/db/backends/util.py", line 40, in execute
    return self.cursor.execute(sql, params)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/django/db/backends/mysql/base.py", line 114, in execute
    return self.cursor.execute(query, args)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/MySQLdb/cursors.py", line 201, in execute
    self.errorhandler(self, exc, value)
  File "/Users/novarg/work/python/py27dj14-env/lib/python2.7/site-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorclass, errorvalue
IntegrityError: (1062, "Duplicate entry '15087' for key 'PRIMARY'")
```

**Comment by EmilStenstrom** [comment]
Hi! We've recently also started to see this error. For us it started appearing when we moved to multiple servers that all run their own processes. We've also using READ-COMMITTED. Reading the MySQL documentation on Deadlocks it seems that client code should always consider this possibility, and retry the same query if this occurs. 

> Always be prepared to re-issue a transaction if it fails due to deadlock. Deadlocks are not dangerous. Just try again. 

Source: http://dev.mysql.com/doc/refman/5.7/en/innodb-deadlocks-handling.html

**My suggestion for this would be to implement retries in Celery to counter this.**
